keps/0020-manual-plan-execution.md

```diff
@@ -173,8 +173,55 @@ Does not look like people are really doing this that much (they usually rather f
 
 #### As a stateful service operator (person) I want to be able to run backup on my stateful service when a backup plan is defined.
 
-### Implementation Details/Notes/Constraints [optional]
+### Implementation Details/Notes/Constraints
 
-TODO: choose a solution
+We've decided to go with `Have a separate CRD just for maintaining commands to execute a plan` solution.
+
+#### Different or unified ways to trigger a plan
+
+Right now, triggering plan executions is fully encapsulated in the instance_controller. It watches updates on Instance objects and decide if it needs to run plan and more importantly which plan (do we run upgrade plan or deploy plan for example?). With manually executed plans, we'll theoretically have two ways to trigger plan:
+
+1. *Non-explicit executions* that are triggered by changing the state of instance (either updating parameter or updating OperatorVersion referencing it - upgrade)
+1. *Explicit executions* (I as operator want to run plan - e.g. backup - right now)
+
+There are multiple approaches how these two can co-exist.
+
+##### PlanExecutionRequest CRD only for external requests
+
+In this solution, PlanExecutionRequest CR is created only for *explicit* executions. With the following rules applied:
+
+If someone submits *explicit execution* (updates the declarative state of the Instance CR)
+
+- the request is **accepted** if (guarded by webhook)
+    - no plan is in progress (already exists in the webhook)
+    - no request for this instance exists which does not have status set (means controller did not pick it up yet)
+
+If someone submits *non-explicit execution* (creates new PlanExecutionRequest CR)
+
+- the request is **accepted** if (guarded by webhook)
+    - no plan is in progress (has to check Instance.Status so a status of a different CRD)
+    - no request for this instance exists which does not have status set (means controller did not pick it up yet)
+
+Also note that `PlanExecutionRequest` here is not a real queue, there can be for simplicitly only one request at a time, all others will be rejected (if not run with force=true). But that is pretty much aligned with the fact that we can run only one plan in parallel.
+
+These two can still race, as these are two separate CRDs, they can be modified concurrently so it's possible that at the same time, update to the instance is accepted as well as request. In that case, instance_controller seeing this during reconciliation can easily change the PlanExecutionRequest.Status to Rejected and pursue with executing the non-explicit trigger as that should always have biggest priority.
+
+##### PlanExecutionRequest is the only way to trigger a plan
```

Comment:
Would be another option to allow updates to `Instance.Spec` which would then trigger a generation of a `PlanExecutionRequest`? Or did you discuss that option?

Reply:
Oh, never mind. You mentioned that the update to `Instance.Spec` would affect all subsequent plan executions without regard to when the update to `Instance.Spec` was made. Nevermind.